File: February2022/Python/SubarraySumEqualsK.py
# TC : O(N)
# SC : O(N)

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        sumOccurenceMap = {0: 1}
        prefixSum, count = 0, 0
        for ind in range(len(nums)):
            prefixSum += nums[ind]
            if prefixSum-k in sumOccurenceMap:
                count += sumOccurenceMap[prefixSum-k]
            if prefixSum not in sumOccurenceMap:
                sumOccurenceMap[prefixSum] = 0
            sumOccurenceMap[prefixSum] += 1
        return count

File: February2022/C++/Subarray_sum_equals_k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> dict;
        int sum=0,res=0;
        dict[0]++;
        for(auto& i:nums)
        {
            sum+=i;
            if(dict.count(sum-k)) res+=dict[sum-k];
            dict[sum]++;
        }
        return res;
    }
};